pandas/io/rdata/_rdata.pyx

```diff
@@ -0,0 +1,437 @@
+        }
+
+        err = RDATA_OK
+        while err == RDATA_OK:
+            err = rdata_set_table_handler(self.rparser, handle_table)
+            err = rdata_set_dim_handler(self.rparser, handle_dim)
+            err = rdata_set_column_handler(self.rparser, handle_column)
+            err = rdata_set_text_value_handler(self.rparser, handle_text_value)
+            err = rdata_set_value_label_handler(self.rparser, handle_value_label)
+            err = rdata_set_column_name_handler(self.rparser, handle_column_name)
+            err = rdata_set_row_name_handler(self.rparser, handle_row_name)
+            err = rdata_set_dim_name_handler(self.rparser, handle_dim_name)
+
+            err = rdata_parse(self.rparser, rfile, <void *>self)
+            rdata_parser_free(self.rparser)
+            break
+
+        if err != RDATA_OK:
+            msg = rdata_error_message(err)
+            raise LibrdataParserError(msg)
+
+        return self.rvalues
+
+
+class LibrdataWriterError(Exception):
+    """
+    Base error class to capture exceptions in librdata writing.
+    """
+    pass
+
+
+cdef ssize_t write_data(const void *bytes, size_t len, void *ctx):
+    cdef int fd = (<int*>ctx)[0]
+
+    IF UNAME_SYSNAME == "Windows":
+        result = _write(fd, bytes, len)
+    ELSE:
+        result = write(fd, bytes, len)
+
+    return result
+
+cdef class LibrdataWriter():
+    """
+    Base class to write RData files.
+
+    Class interfaces with librdata C library to iterate through dictionaries
+    of each DataFrame column according to correspoinding dtype.
+    Single callback above is usedd in exposed `init`` method.
+    """
+    cdef:
+        int fd
+        int row_count
+        dict rdict
+        dict rformats
+        dict rtypes
+        bytes file_name
+        bytes tbl_name
+        rdata_writer_t *writer
+        rdata_column_t *py_col
+
+    cdef write_col_data(self, i, kdata, vdata, ktype, vtype):
+        py_col = rdata_get_column(self.writer, i)
+        rdata_begin_column(self.writer, py_col, self.row_count)
+
+        if vtype == "bool":
+            for k, v in vdata.items():
+                rdata_append_logical_value(self.writer, v)
+
+        if vtype.startswith(("int", "uint")):
+            for k, v in vdata.items():
+                rdata_append_int32_value(self.writer, v)
+
+        if vtype.startswith("float"):
+            for k, v in vdata.items():
+                rdata_append_real_value(self.writer, v)
+
+        if vtype.startswith("datetime64"):
+            for k, v in vdata.items():
+                rdata_append_timestamp_value(self.writer, v)
+
+        if vtype == "object":
+            for k, v in vdata.items():
+                if v == v:
+                    rdata_append_string_value(self.writer, v)
+                else:
+                    rdata_append_string_value(self.writer, NULL)
+
+        rdata_end_column(self.writer, py_col)
+
+    cpdef write_rdata(self, rfile, rdict, rformat, tbl_name=None):
```

Comment:
Why not plain `def`?

Reply:
Learning as I was doing. Either way can work.